File: oxidize-python/oxidize_python/core/model/mtp.py
"""MTP generation mirroring oxidize-golang/core/model/mtp.go."""

from __future__ import annotations

from oxidize_python.core.ggufcore import gguf as ggufcore
from oxidize_python.core.model.generation import (
    ERR_GENERATION_FINISHED,
    GenerationConfig,
    GenerationError,
)
from oxidize_python.core.model.model import Model, Session, Token
from oxidize_python.core.model.sampling import sample


def has_mtp_weights(path: str) -> bool:
    try:
        mapped = ggufcore.load_mapped(path)
    except OSError:
        return False
    for tensor in mapped.parsed.tensor_infos:
        name = tensor.name.lower()
        if "nextn" in name or "mtp" in name:
            return True
    return False


class MtpGenerationStream:
    def __init__(self, model: Model, session: Session, config: GenerationConfig) -> None:
        self.model = model
        self.session = session
        self.config = config
        self.done = False
        self.prompt: list[Token] = []

    def seed(self, prompt: list[Token]) -> None:
        self.prompt = list(prompt)

    def next(self) -> tuple[Token, bool, GenerationError | None]:
        if self.done:
            return 0, True, ERR_GENERATION_FINISHED
        context_tokens = list(self.prompt)
        logits = self.model.forward(context_tokens, self.session)
        token = sample(logits, self.config.sampling, None)
        if token == self.config.stop_token:
            self.done = True
            return token, True, None
        self.prompt.append(token)
        if len(self.prompt) >= self.config.max_new_tokens:
            self.done = True
        return token, self.done, None

File: oxidize-golang/core/model/mtp.go
package model

import (
	"context"
	"strings"

	"github.com/Zapdev-labs/oxidize/golang/core/ggufcore"
)

// HasMTPWeights reports whether a GGUF file contains MTP/nextn tensors.
func HasMTPWeights(path string) bool {
	mapped, err := ggufcore.LoadMapped(path)
	if err != nil {
		return false
	}
	for _, t := range mapped.Parsed.TensorInfos {
		n := strings.ToLower(t.Name)
		if strings.Contains(n, "nextn") || strings.Contains(n, "mtp") {
			return true
		}
	}
	return false
}

// MtpGenerationStream uses in-GGUF MTP heads for multi-token draft steps.
type MtpGenerationStream struct {
	model   Model
	session *Session
	config  GenerationConfig
	done    bool
	prompt  []Token
}

// NewMtpGenerationStream constructs an MTP-backed generation stream.
func NewMtpGenerationStream(model Model, session *Session, config GenerationConfig) *MtpGenerationStream {
	return &MtpGenerationStream{model: model, session: session, config: config}
}

// Seed sets the prompt tokens.
func (s *MtpGenerationStream) Seed(prompt []Token) {
	s.prompt = append([]Token(nil), prompt...)
}

// Next generates the next token (MTP-aware path uses the same forward as baseline today).
func (s *MtpGenerationStream) Next(ctx context.Context) (Token, bool, error) {
	if s.done {
		return 0, true, errGenerationFinished
	}
	if err := ctx.Err(); err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}
	contextTokens := append([]Token(nil), s.prompt...)
	logits, err := s.model.Forward(contextTokens, s.session)
	if err != nil {
		return 0, true, &GenerationError{Message: err.Error()}
	}
	token, err := Sample(logits, s.config.Sampling, nil)
	if err != nil {
		return 0, true, err
	}
	if token == s.config.StopToken {
		s.done = true
		return token, true, nil
	}
	s.prompt = append(s.prompt, token)
	if len(s.prompt) >= s.config.MaxNewTokens {
		s.done = true
	}
	return token, s.done, nil
}
